RearrangingCars.java

```diff
@@ -0,0 +1,207 @@
+package CodeU_Assignment6;
+
+import java.util.ArrayList;
+import java.util.HashSet;
+
+public class RearrangingCars {
+    /** rearrangingSteps(initial, expected) returns the steps required to arrange the cars
+     *  according to the "expected" rearrangement, starting from the "initial" rearrangement. 
+     *  
+     *  Solution:
+     *      1. Check if the current and the expected rearrangements are valid, if not, then throw an exception.
+     *      2. If the arrays are of length 0 or are exactly equal, there is no steps to be taken, hence return "".
+     *      3. Otherwise, call rearrangingStepsHelper() to get the required series of steps. 
+     *         Return the required steps as a String.           
+     * 
+     * @param initial: an array representing initial rearrangement of the cars
+     * @param expected: an array representing expected rearrangement of the cars
+     * @return String: A string containing a series of steps required to arrange the cars
+     *  according to the "expected" rearrangement, starting from the "initial" rearrangement.
+     * @throws RearrangementException 
+     */
+    public static String rearrangingSteps(int [] initial, int [] expected) throws RearrangementExceptions{
+        
+        StringBuilder str = new StringBuilder();
+        
+        //If the initial or expected rearrangement array is null, then throw an exception.
+        if(initial == null || expected == null){
+            throw new NullRearrangementsException("Either the initial rearrangement or the expected rearrangement is null");
+        }
+        
+        //If the initial and expected rearrangement array do not have same length, then throw an exception.
+        else if(initial.length != expected.length){
+            throw new LengthMismatchException("The length of arrays representing the initial and expected rearrangement are not same");
+        }
+        
+        //If the initial or expected rearrangement array is not a valid permutation of {0,...,array.length-1}, then throw an exception.
+        else if(!isValidPermutation(initial) || !isValidPermutation(expected)){
+            throw new InvalidRearrangementException("The initial or the expected rearrangement is invalid!");
+        }
+        
+        else{
+            //If the initial and expected rearrangement is empty (no slots), then nothing to do, simply return "".
+            if(initial.length == 0){

```

Comment:
Oh wait, I think I didn't get you. To reply to your comment "I think lines from 43-56 are *not* necessary as your algorithm will quickly detect that arrays are empty or equal"...
I said it is necessary to put those lines as it a scan would take n time, whereas if I do not check (pre-process) it, then it might create an inverseArray which would take n space also along with n time. So it is better to pre-process it. We can discuss more in the evening meeting.